File: frontend/lib/api/tree.ts
/**
 * Tree API
 * 
 * Handles fetching directory tree structure.
 * RESTful v1 API path: GET /api/v1/tree
 */

import { apiClient } from './client';
import type { TreeNode } from '@/types';

interface GetTreeParams {
  dir?: string;
}

interface TreeResponse {
  root: TreeNode;
  currentDir: string;
}

/**
 * Get directory tree structure
 * GET /api/v1/tree?dir={dir}
 */
export async function getTree(params: GetTreeParams = {}): Promise<TreeResponse> {
  const query = params.dir ? `?dir=${encodeURIComponent(params.dir)}` : '';
  return apiClient.get<TreeResponse>(`/api/v1/tree${query}`);
}

File: internal/handler/tree.go
// Package handler provides HTTP request handlers for the application.
package handler

import (
	"net/http"

	"terminalog/internal/model"
	"terminalog/internal/service"
	"terminalog/pkg/utils"
)

// TreeHandler handles directory tree-related HTTP requests.
type TreeHandler struct {
	svc *service.ArticleService
}

// NewTreeHandler creates a new TreeHandler instance.
func NewTreeHandler(svc *service.ArticleService) *TreeHandler {
	return &TreeHandler{svc: svc}
}

// Get handles GET /api/v1/tree.
func (h *TreeHandler) Get(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	// Get directory parameter (optional)
	dir := r.URL.Query().Get("dir")

	// Get tree
	tree, err := h.svc.GetTree(ctx, dir)
	if err != nil {
		// If the directory doesn't exist, return root tree
		if dir != "" {
			tree, err = h.svc.GetTree(ctx, "")
			if err != nil {
				utils.RespondInternalServerError(w, err.Error())
				return
			}
		} else {
			utils.RespondInternalServerError(w, err.Error())
			return
		}
	}

	// Respond
	utils.RespondJSON(w, http.StatusOK, model.TreeResponse{
		Root:       tree,
		CurrentDir: dir,
	})
}
